### PR TITLE
HuffmanTree instead of iterating through Edges

### PR DESCRIPTION
**Pull request by Majklikikik**
Added `Gnode.findHoppingDestination()`
 to Gnode to replace findHoppingDestination in kmccalculator. The new algorith makes a huffmantree and searches quicker for systems with many connections per node

### CODE REVIEW

File: include/votca/xtp/gnode.h [hunk 1]
```diff
@@ -16,19 +16,48 @@
...
 
 
+using namespace std;
+
 
 namespace votca { namespace xtp {
 
+
+//hnode as huffman-node
+struct hnode{
```
Comment (JensWehner) [inline include/votca/xtp/gnode.h]: Rename huffman-node, maybe add wiki link

File: include/votca/xtp/gnode.h [hunk 2]
```diff
@@ -16,19 +16,48 @@
...
+    //id's of the child nodes
+    //if node is a leaf, both are null
+    //hnode * small;
```
Comment (JensWehner) [inline include/votca/xtp/gnode.h]: remove commented out code

File: include/votca/xtp/gnode.h [hunk 3]
```diff
@@ -16,19 +16,48 @@
...
+    //hnode * big;
+
+    //own index/id
```
Comment (JensWehner) [inline include/votca/xtp/gnode.h]: unnecessary comment, id is enough

File: include/votca/xtp/gnode.h [hunk 4]
```diff
@@ -16,19 +16,48 @@
...
+    int id;
+    //id's (indexes) of the children
+    int leftId;
```
Comment (JensWehner) [inline include/votca/xtp/gnode.h]: leftchild_id as a name is better than comments

File: include/votca/xtp/gnode.h [hunk 5]
```diff
@@ -16,19 +16,48 @@
...
+    int rightId;
+    //probability of this node being chosen
+    double prob;
+    //pointer to the gnode, if hnode is a leaf
+    //GLink * edge; not needed, as the index in the edge list is the same as the hnodes index
```
Comment (JensWehner) [inline include/votca/xtp/gnode.h]: remove commented out code

File: include/votca/xtp/gnode.h [hunk 6]
```diff
@@ -16,19 +16,48 @@
...
+    //in the htree vector
+};
+
+
+
 class GNode
 {
+    private:
+        void organizeProbabilities(int id, double add);
```
Comment (JensWehner) [inline include/votca/xtp/gnode.h]: move private to the bottom of file

File: include/votca/xtp/gnode.h [hunk 7]
```diff
@@ -46,13 +75,17 @@ class GNode
...
-    
         void AddEvent(int seg2, double rate12, tools::vec dr, double Jeff2, double reorg_out);
         const double &getEscapeRate(){return escape_rate;}
         void InitEscapeRate();
         void AddDecayEvent(double decayrate);
         void ReadfromSegment(ctp::Segment* seg, int carriertype);
         void AddEventfromQmPair(ctp::QMPair* pair,int carriertype);
+        
+        GLink* findHoppingDestination(double p);
+        hnode * root;
```
Comment (JensWehner) [inline include/votca/xtp/gnode.h]: why keep a pointer to the root of the tree? is that not simply htree[0]?

File: src/libxtp/gnode.cc [hunk 8]
```diff
@@ -11,17 +11,33 @@
...
 
 using namespace std;
 
+
+
+
 namespace votca {
     namespace xtp {
+
+        struct Comparer
```
Comment (JensWehner) [inline src/libxtp/gnode.cc]: this can be a simple lambda function not its own class

File: src/libxtp/gnode.cc [hunk 9]
```diff
@@ -61,6 +77,77 @@ void GNode::InitEscapeRate()
...
+    while (node->leftId!=-1){
+        if (p>node->prob) node=&(htree[node->leftId]);
+        else node=&(htree[node->rightId]);
+    }
+    return &events[node->id];
+}
+
+void GNode::MakeHuffTree(){
+    //queue of the nodes, sorted by probability
+    priority_queue<hnode *,vector<hnode *>,Comparer> queue;
```
Comment (JensWehner) [inline src/libxtp/gnode.cc]: lambda function

File: src/libxtp/gnode.cc [hunk 10]
```diff
@@ -61,6 +77,77 @@ void GNode::InitEscapeRate()
...
+    htree=vector<hnode>(events.size()*2-1);
```
Comment (JensWehner) [inline src/libxtp/gnode.cc]: why is htree twice as large as events.size()

File: src/libxtp/gnode.cc [hunk 11]
```diff
@@ -61,6 +77,77 @@ void GNode::InitEscapeRate()
...
+    //first, make a leaf for every GLink
+    int index=0;
+    for (GLink L:events){
```
Comment (JensWehner) [inline src/libxtp/gnode.cc]: if you use an index anyway use `for unsigned index=0;index<events.size`

Comment (JensWehner) [reply src/libxtp/gnode.cc]: but probably going from indices to pointers is the better idea

File: src/libxtp/gnode.cc [hunk 12]
```diff
@@ -61,6 +77,77 @@ void GNode::InitEscapeRate()
...
+        htree[index].leftId=-1;
+        htree[index].rightId=-1;
+        htree[index].id=index;
+        queue.push(&(htree[index]));
+        index++;
+    }
+
+     //now connect the hnodes, making a new one for every connection:
+     //always take the two nodes with the smallest probability and "combine" them, repeat, until just one node (the root) is left.
+     hnode * h1;
```
Comment (JensWehner) [inline src/libxtp/gnode.cc]: move the pointers inside the while loop, they are not used outside

File: src/libxtp/gnode.cc [hunk 13]
```diff
@@ -61,6 +77,77 @@ void GNode::InitEscapeRate()
...
+         queue.pop();
+         htree[index].prob=h1->prob+h2->prob;
+         htree[index].leftId=h1->id;
+         htree[index].rightId=h2->id;
+         htree[index].id=index;
+         queue.push(&(htree[index]));
+         index++;
+     }
+     //save the root
+     root=&(htree[index-1]);
```
Comment (JensWehner) [inline src/libxtp/gnode.cc]: pointer instead of index

File: src/libxtp/gnode.cc [hunk 14]
```diff
@@ -61,6 +77,77 @@ void GNode::InitEscapeRate()
...
+    moveProbabilities(index-1);
+}
+
+void GNode::organizeProbabilities(int id,double add){
+
+    //adds "add" to the probability, then calls itself recursively.
+    //this calculates the probabilities needed to traverse the tree quickly
+    htree[id].prob+=add;
+    //if leftId=-1 (=> node is leaf), returns
+    if (htree[id].leftId==-1) return;
```
Comment (JensWehner) [inline src/libxtp/gnode.cc]: use own boolean

File: src/libxtp/gnode.cc [hunk 15]
```diff
@@ -61,6 +77,77 @@ void GNode::InitEscapeRate()
...
+}
+
+void GNode::moveProbabilities(int id){
+    //moves the probabilities "one up" so that htree[id].prob can be checked instead of htree[htree[id].rightId].prob
+    if (htree[id].rightId!=-1){
+        htree[id].prob=htree[htree[id].rightId].prob;
+        moveProbabilities(htree[id].rightId);
+        moveProbabilities(htree[id].leftId);
+    }
+    else htree[id].prob=-1;
```
Comment (JensWehner) [inline src/libxtp/gnode.cc]: why -1?

File: src/libxtp/kmccalculator.cc [hunk 16]
```diff
@@ -111,6 +110,7 @@ namespace votca {
 
         for (unsigned int i = 0; i < _nodes.size(); i++) {
             _nodes[i]->InitEscapeRate();
```
Comment (JensWehner) [inline src/libxtp/kmccalculator.cc]: make c++11 iteration

File: src/libxtp/kmccalculator.cc [hunk 17]
```diff
@@ -325,15 +325,15 @@ namespace votca {
...
-                
+               
 
             }
-            
-            // Initialise escape rates
+             // Initialise escape rates
                 for (auto* node:_nodes) {
                     node->InitEscapeRate();
+                    node->MakeHuffTree();
```
Comment (JensWehner) [inline src/libxtp/kmccalculator.cc]: as here;

File: src/libxtp/kmccalculator.cc [hunk 18]
```diff
@@ -362,15 +362,7 @@ namespace votca {
         
         GLink* KMCCalculator::ChooseHoppingDest(GNode* node){
             double u = 1 - _RandomVariable.rand_uniform();
```
Comment (JensWehner) [inline src/libxtp/kmccalculator.cc]: nice change

File: include/votca/xtp/huffmantree.h [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+class huffmanTree
+{
+private:
+    void organizeProbabilities(huffmanNode *n, double add);
+    void moveProbabilities(huffmanNode *n);
+    vector <huffmanNode> htree;
+    bool treeIsMade=false;
+public:
+    double escape_rate=0.0;
+    vector <GLink> * events=nullptr;
```
Comment (JensWehner) [inline include/votca/xtp/huffmantree.h]: no public variables please use setter and getter if necessary

File: src/libxtp/huffmantree.cc [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+#include <queue>
+#include <votca/xtp/huffmantree.h>
+#include <votca/xtp/glink.h>
+using namespace std;
+
+
+
+namespace votca {namespace xtp {
+
+huffmanTree::huffmanTree()
```
Comment (JensWehner) [inline src/libxtp/huffmantree.cc]: C++ writes automatically a default constructor so remove this.

File: src/libxtp/huffmantree.cc [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+{
+}
+
+//https://en.wikipedia.org/wiki/Huffman_coding
+void huffmanTree::makeTree(){
+    if (!events) throw runtime_error("Error in Huffmantree::makeTree : Pointer to Events not set!");
+
+    //queue of the nodes, sorted by probability
+    auto compare = [](huffmanNode * n1, huffmanNode * n2)
```
Comment (JensWehner) [inline src/libxtp/huffmantree.cc]: good

File: src/libxtp/huffmantree.cc [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+namespace votca {namespace xtp {
+
+huffmanTree::huffmanTree()
+{
+}
+
+//https://en.wikipedia.org/wiki/Huffman_coding
+void huffmanTree::makeTree(){
+    if (!events) throw runtime_error("Error in Huffmantree::makeTree : Pointer to Events not set!");
```
Comment (JensWehner) [inline src/libxtp/huffmantree.cc]: good

File: src/libxtp/huffmantree.cc [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+    auto compare = [](huffmanNode * n1, huffmanNode * n2)
+    { return n1->probability>n2->probability;};
+    priority_queue<huffmanNode *,vector<huffmanNode *>, decltype(compare)> queue(compare);
+
+    htree=vector<huffmanNode>(events->size()%2?events->size():events->size()-1);
+
+    auto comp2 = [](GLink * e1, GLink * e2){
+        return e1->rate>e2->rate;
+    };
+    priority_queue<GLink *, vector<GLink *>, decltype(comp2)> eventQueue(comp2);
```
Comment (JensWehner) [inline src/libxtp/huffmantree.cc]: put a comment in why priority queue, it is a rather seldomly used structure

File: src/libxtp/huffmantree.cc [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+
+    int i=0;
```
Comment (JensWehner) [inline src/libxtp/huffmantree.cc]: can you remove this counter, and use some other iteration, if not give it a more explicit name

File: src/libxtp/huffmantree.cc [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+    //reorganize the probabilities: in every node, add the probability of one subtree ("small")
+    //to all nodes of the other subtree.
+    organizeProbabilities(&htree[htree.size()-1],0);
+    moveProbabilities(&htree[htree.size()-1]);
+    treeIsMade=true;
+}
+
+
+
+void huffmanTree::organizeProbabilities(huffmanNode * n,double add){
```
Comment (JensWehner) [inline src/libxtp/huffmantree.cc]: organize is not specific enough, do you mean allglomerate, sum or what? Find a better function name

File: src/libxtp/huffmantree.cc [hunk 26]
```diff
@@ -0,0 +1,119 @@
...
+        return;
+    }
+
+    organizeProbabilities(n->leftChild,add+n->rightChild->probability);
+    organizeProbabilities(n->rightChild,add);
+}
+
+
+
+void huffmanTree::moveProbabilities(huffmanNode * n){
```
Comment (JensWehner) [inline src/libxtp/huffmantree.cc]: moveProbabilties can also be more explicit

File: src/libxtp/huffmantree.cc [hunk 27]
```diff
@@ -0,0 +1,119 @@
...
+        moveProbabilities(n->rightChild);
+        moveProbabilities(n->leftChild);
+    }
+}
+
+
+
+GLink * huffmanTree::findHoppingDestination(double p){
+    if (!treeIsMade) throw runtime_error("Tried to find Hopping Destination without initializing the Huffmantree first!");
+    huffmanNode * node=&(htree[htree.size()-1]);
```
Comment (JensWehner) [inline src/libxtp/huffmantree.cc]: htree.back() is the last element, so please replace

**Comment by junghans** [comment]
@JensWehner template parameter?

**Comment by JensWehner** [comment]
yes good idea

**Pull request merged**